File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/ReplanFromHere.java
package net.cyclestreets.liveride;

import net.cyclestreets.CycleStreetsPreferences;
import net.cyclestreets.routing.Journey;
import net.cyclestreets.routing.Route;
import net.cyclestreets.routing.Waypoints;

import org.osmdroid.api.IGeoPoint;
import org.osmdroid.util.GeoPoint;

final class ReplanFromHere extends LiveRideState implements Route.Listener
{
  private LiveRideState next_;

  ReplanFromHere(final LiveRideState previous, final GeoPoint whereIam) {
    super(previous);
    notify("Too far away. Re-planning the journey.");

    next_ = this;

    final IGeoPoint finish = Route.waypoints().last();
    Route.softRegisterListener(this);
    Route.PlotRoute(CycleStreetsPreferences.routeType(),
                    CycleStreetsPreferences.speed(),
                    getContext(),
                    Waypoints.fromTo(whereIam, finish));
  }

  @Override
  public LiveRideState update(Journey journey, GeoPoint whereIam, int accuracy) {
    return next_;
  }

  @Override
  public boolean isStopped() { return false; }

  @Override
  public boolean arePedalling() { return true; }

  @Override
  public void onNewJourney(Journey journey, Waypoints waypoints) {
    next_ = new HuntForSegment(this);
    Route.unregisterListener(this);
  }

  @Override
  public void onResetJourney() {
  }
}

File: libraries/cyclestreets-view/src/main/java/net/cyclestreets/liveride/ReplanFromHere.kt
package net.cyclestreets.liveride

import android.util.Log
import net.cyclestreets.CycleStreetsPreferences
import net.cyclestreets.routing.Journey
import net.cyclestreets.routing.Route
import net.cyclestreets.routing.Segment
import net.cyclestreets.routing.Waypoints
import net.cyclestreets.util.Logging

import org.osmdroid.util.GeoPoint

private val TAG = Logging.getTag(ReplanFromHere::class.java)

internal class ReplanFromHere(previous: LiveRideState, whereIam: GeoPoint) : LiveRideState(previous), Route.Listener {
    private var next: LiveRideState? = null

    init {
        notify("Too far away. Re-planning the journey.")

        next = this

        val activeSegment = Route.journey().activeSegment()
        val remainingWaypoints: Waypoints = when (activeSegment) {
            is Segment.Start ->  Route.waypoints().startingWith(whereIam)
            is Segment.End -> Waypoints.fromTo(whereIam, Route.waypoints().last())
            is Segment.Waymark -> Route.waypoints().fromLeg(activeSegment.legNumber()).startingWith(whereIam)
            is Segment.Step -> Route.waypoints().fromLeg(activeSegment.legNumber()).startingWith(whereIam)
            else -> {
                Log.w(TAG, "Unexpected segment type ${activeSegment?.javaClass ?: "'null'"}")
                throw IllegalStateException("Unexpected segment type ${activeSegment?.javaClass ?: "'null'"}")
            }
        }
        Route.softRegisterListener(this)
        Route.LiveReplanRoute(CycleStreetsPreferences.routeType(),
                              CycleStreetsPreferences.speed(),
                              context,
                              remainingWaypoints)
    }

    override fun update(journey: Journey, myLocation: GeoPoint, accuracy: Int): LiveRideState {
        return next!!
    }

    override fun isStopped(): Boolean { return false }
    override fun arePedalling(): Boolean { return true }

    override fun onNewJourney(journey: Journey, waypoints: Waypoints) {
        next = HuntForSegment(this)
        Route.unregisterListener(this)
    }

    override fun onResetJourney() {}
}
